include/raptor/prepare/cutoff.hpp

```diff
@@ -0,0 +1,129 @@
+// shipped with this file and also available at: https://github.com/seqan/raptor/blob/main/LICENSE.md
+// --------------------------------------------------------------------------------------------------
+
+#pragma once
+
+#include <seqan3/io/sequence_file/format_fasta.hpp>
+
+#include <raptor/argument_parsing/prepare_arguments.hpp>
+
+namespace raptor
+{
+
+class cutoff
+{
+public:
+    cutoff() = default;
+    cutoff(cutoff const &) = default;
+    cutoff & operator=(cutoff const &) = default;
+    cutoff(cutoff &&) = default;
+    cutoff & operator=(cutoff &&) = default;
+    ~cutoff() = default;
+
+    cutoff(prepare_arguments const & arguments) : fixed_cutoff{arguments.kmer_count_cutoff}
+    {
+        if (arguments.use_filesize_dependent_cutoff)
+        {
+            cutoff_kind = cutoff_kinds::filesize;
+        }
+        else
+        {
+            cutoff_kind = cutoff_kinds::fixed;
+        }
+    }
+
+    size_t get(std::filesystem::path const & filename) const noexcept
+    {
+        switch (cutoff_kind)
+        {
+        case cutoff_kinds::filesize:
+            return impl(filename);
+        default:
+        {
+            assert(cutoff_kind == cutoff_kinds::fixed);
+            return fixed_cutoff;
+        }
+        }
+    }
+
+    static inline bool file_is_compressed(std::filesystem::path const & filepath)
+    {
+        std::filesystem::path const extension = filepath.extension();
+        return extension == ".gz" || extension == ".bgzf" || extension == ".bz2";
+    }
+
+private:
+    enum class cutoff_kinds
+    {
+        fixed,
+        filesize
+    };
+
+    cutoff_kinds cutoff_kind{cutoff_kinds::fixed};
+
+    uint16_t fixed_cutoff{};
+    // Cutoffs and bounds from Mantis
+    // Mantis ignores k-mers which appear less than a certain cutoff. The cutoff is based on the file size of a
+    // gzipped fastq file. Small files have only a cutoff of 1 while big files have a cutoff value of 50.
+    static constexpr std::array<uint16_t, 4> const cutoffs{1u, 3u, 10u, 20u};
+    static constexpr std::array<uint64_t, 4> const cutoff_bounds{314'572'800ULL,
+                                                                 524'288'000ULL,
+                                                                 1'073'741'824ULL,
+                                                                 3'221'225'472ULL};
+
+    uint16_t impl(std::filesystem::path const & filename) const
+    {
+        bool const is_compressed = file_is_compressed(filename);
+        bool const is_fasta = check_for_fasta_format(filename);
+
+        // Since the curoffs are based on the filesize of a gzipped fastq file, we try account for the other cases:
+        // We multiply by two if we have fasta input.
+        // We divide by 3 if the input is not compressed.
+        size_t const filesize = std::filesystem::file_size(filename) * (is_fasta ? 2 : 1) / (is_compressed ? 1 : 3);
```

Comment:
same name as the `cutoff_kind::filesize`. Maybe the kind could be `cutoff_kind::filesize_dependent`?